pkg/querier/tripperware/shard_by.go

```diff
@@ -0,0 +1,94 @@
+package tripperware
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/thanos-io/thanos/pkg/querysharding"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"github.com/weaveworks/common/httpgrpc"
+
+	cquerysharding "github.com/cortexproject/cortex/pkg/querysharding"
+	"github.com/cortexproject/cortex/pkg/tenant"
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+func ShardByMiddleware(logger log.Logger, limits Limits, merger Merger) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return shardBy{
+			next:   next,
+			limits: limits,
+			merger: merger,
+			logger: logger,
+		}
+	})
+}
+
+type shardBy struct {
+	next          Handler
+	limits        Limits
+	logger        log.Logger
+	merger        Merger
+	queryAnalyzer *querysharding.QueryAnalyzer
+}
+
+func (s shardBy) Do(ctx context.Context, r Request) (Response, error) {
+	tenantIDs, err := tenant.TenantIDs(ctx)
+
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+
+	numShards := validation.SmallestPositiveIntPerTenant(tenantIDs, s.limits.QueryVerticalShardSize)
```

Comment:
I think its documented on the function itself:

```
// SmallestPositiveIntPerTenant is returning the minimal positive value of the
// supplied limit function for all given tenants.
func SmallestPositiveIntPerTenant(tenantIDs []string, f func(string) int) int {
```

Reply:
Right, for the code level it's clear. But if I am an user that don't know the code base well, and I enable federated query + vertical query sharding, I think it would help that the flag documentation tells me that the effective value is the minimum of all the configured  limit for each tenant.

Reply:
Make sense.. But this is used on multiples APIs..

I will create a follow up PR to document this behavior on all configs.